File: dotnet/Lexy/Lexy.Compiler/Language/Expressions/IdentifierExpression.cs
using System.Collections.Generic;
using Lexy.Compiler.Language.Types;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language.Expressions
{
    public class IdentifierExpression : Expression
    {
        public VariableSource VariableSource { get; private set; }

        public string Identifier { get; }

        private IdentifierExpression(string identifier, ExpressionSource source, SourceReference reference) : base(source, reference)
        {
            Identifier = identifier;
        }

        public static ParseExpressionResult Parse(ExpressionSource source)
        {
            var tokens = source.Tokens;
            if (!IsValid(tokens))
            {
                return ParseExpressionResult.Invalid<IdentifierExpression>("Invalid expression");
            }

            var variableName = tokens.TokenValue(0);
            var reference = source.CreateReference();

            var expression = new IdentifierExpression(variableName, source, reference);

            return ParseExpressionResult.Success(expression);
        }

        public static bool IsValid(TokenList tokens)
        {
            return tokens.Length == 1
                   && tokens.IsTokenType<StringLiteralToken>(0);
        }

        public override IEnumerable<INode> GetChildren()
        {
            yield break;
        }

        protected override void Validate(IValidationContext context)
        {
            context.VariableContext.EnsureVariableExists(Reference, Identifier);

            var variableSource = context.VariableContext.GetVariableSource(Identifier);
            if (variableSource == null)
            {
                context.Logger.Fail(Reference, "Can't define source of variable: " + Identifier);
                return;
            }
            VariableSource = variableSource.Value;
        }

        public override VariableType DeriveType(IValidationContext context) =>
            context.VariableContext.GetVariableType(Identifier);
    }

}

File: typescipt/src/language/expressions/identifierExpression.ts


export class IdentifierExpression extends Expression {
   public VariableSource VariableSource { get; private set; }

   public string Identifier

   private IdentifierExpression(string identifier, ExpressionSource source, SourceReference reference) : base(source,
     reference) {
     Identifier = identifier;
   }

   public static parse(source: ExpressionSource): ParseExpressionResult {
     let tokens = source.Tokens;
     if (!IsValid(tokens)) return ParseExpressionResult.Invalid<IdentifierExpression>(`Invalid expression`);

     let variableName = tokens.TokenValue(0);
     let reference = source.CreateReference();

     let expression = new IdentifierExpression(variableName, source, reference);

     return ParseExpressionResult.Success(expression);
   }

   public static isValid(tokens: TokenList): boolean {
     return tokens.Length == 1
        && tokens.IsTokenType<StringLiteralToken>(0);
   }

   public override getChildren(): Array<INode> {
     yield break;
   }

   protected override validate(context: IValidationContext): void {
     if (!context.VariableContext.EnsureVariableExists(Reference, Identifier)) return;

     let variableSource = context.VariableContext.GetVariableSource(Identifier);
     if (variableSource == null) {
       context.Logger.Fail(Reference, `Can't define source of variable: ` + Identifier);
       return;
     }

     VariableSource = variableSource.Value;
   }

   public override deriveType(context: IValidationContext): VariableType {
     return context.VariableContext.GetVariableType(Identifier);
   }
}
